### PR TITLE
v1.2.0 - Grand(and greater)parent rollups

### LINKED ISSUE #46: Add support for grandchild to grandparent rollups (and potentially more)
This feature request from @elements:

Currently, rollup assumes that the only `1 : many` relationship is from the parent to each "child" object in the hierarchy - but it would be great to allow something like rolling up fields from Opp Line Items straight to Account, for example; cutting out the Opportunity middleman. 

For multiple "jumps" between objects, the relationship between the "lookup field" (since it can be a string or Id) to the parent object would have to be established (possibly using "Apex map" syntax a la @tsalb's [Custom Column Labels in lwc-utils](https://github.com/tsalb/lwc-utils/wiki/soqlDatatable:-Custom-Column-Labels), or with another CMDT that's related to `Rollup__mdt`). Being able to use filtering criteria on the in-between objects should also be supported.

### PR DESCRIPTION
**Pull request by jamessimone**
* Fixes #46 by adding the ability to perform grandparent rollups!

- grandparent rollups started from the child have no practical limit - you can rollup values to a child's great-great-grandparent, for example
- grandparent rollups started from the parent respect [SOQL's map relationship-field hopping of 5 levels](https://developer.salesforce.com/docs/atlas.en-us.soql_sosl.meta/soql_sosl/sforce_api_calls_soql_relationships_query_limits.htm):

> In each specified relationship, no more than five levels can be specified in a child-to-parent relationship. For example, Contact.Account.Owner.FirstName (three levels)

* Deprecated `RollupControl__mdt.MaximumRollupRetries__c` in favor of `RollupControl__mdt.MaxRollupRetries__c` to keep naming consistent
* Updated query architecture for rollups initiated from parents, in general - bind variables are now used to prevent query size limits from being reached

### CODE REVIEW

**Review by jongpie — COMMENTED** [review]
I did an initial review of the code - things look great overall, I don't have any major changes to your approach for handling grandparents. But, I added a few comments related to naming conventions and a few clarifying questions. I'm planning to do some more testing in a scratch org to see if I find any bugs - I'll let ya know if I find anything else!

File: rollup/main/default/classes/RollupCalculator.cls [hunk 1]
```diff
@@ -681,10 +681,11 @@ public without sharing abstract class RollupCalculator {
     protected override void setReturnValue() {
       String trimmedDelimiter = this.concatDelimiter.trim();
       String possibleReturnValue = this.stringVal.normalizeSpace();
-      String withoutEndingComma = possibleReturnValue.endsWith(trimmedDelimiter)
-        ? possibleReturnValue.substringBeforeLast(trimmedDelimiter)
-        : possibleReturnValue;
-      this.returnVal = (withoutEndingComma.startsWith(trimmedDelimiter) ? withoutEndingComma.substring(1, withoutEndingComma.length()) : withoutEndingComma)
+      while(possibleReturnValue.endsWith(trimmedDelimiter)) {
```
[the remark this answers is not in the document]
Comment (jamessimone) [reply rollup/main/default/classes/RollupCalculator.cls:684]: I found while testing the grandparent rollups that it was possible in a delete concat/concat distinct event for the _value_ to to be removed without the delimiter being removed. This is due to the shared code path between update/delete in the `replaceWithDelimiter` function - I'll have to go back and review whether or not those code paths sharing responsibilities makes sense before resolving this; it could be that delete should just always remove both the value and the delimiter. I have good enough test coverage on deletes that returning to this version of the code and trying that should make the matter clear as day!

File: rollup/main/default/classes/RollupEvaluatorTests.cls [hunk 2]
```diff
@@ -262,7 +259,7 @@ private class RollupEvaluatorTests {
     Opportunity oppThree = new Opportunity(ForecastCategoryName = 'Best Case');
     Opportunity oppFour = new Opportunity(ForecastCategoryName = 'Commit');
     Opportunity oppFive = new Opportunity(Amount = 5, ForecastCategoryName = 'Closed');
-    Formula.recalculateFormulas(new List<Opportunity>{ oppOne, oppTwo, oppThree, oppFour, oppFive });
+    System.Formula.recalculateFormulas(new List<Opportunity>{ oppOne, oppTwo, oppThree, oppFour, oppFive });
```
[the remark this answers is not in the document]
Comment (jamessimone) [reply rollup/main/default/classes/RollupEvaluatorTests.cls:262]: Ha - I actually added this because I was (and am) using the `apex-assist` VS Code plugin, and [there was an issue with its linting of this method](https://github.com/nawforce/apex-link/issues/31) which, while closed, has yet to be released to the plugin. I can revert this - good catch!

File: rollup/main/default/objects/Rollup__mdt/fields/GrandparentRelationshipFieldPath__c.field-meta.xml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<CustomField xmlns="http://soap.sforce.com/2006/04/metadata">
+    <fullName>GrandparentRelationshipFieldPath__c</fullName>
```
[the remark this answers is not in the document]
Comment (jamessimone) [reply rollup/main/default/objects/Rollup__mdt/fields/GrandparentRelationshipFieldPath__c.field-meta.xml:3]: I think I'll stick to it for now, but this is good food for thought in the pre-managed package era. Since the next big thing I want to do is re-do the folder structure, I'll 📍this - I love the idea, overall, and think I'll probably update them long-term

File: rollup/main/default/objects/RollupControl__mdt/fields/MaxParentRowsUpdatedAtOnce__c.field-meta.xml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<CustomField xmlns="http://soap.sforce.com/2006/04/metadata">
+    <fullName>MaxParentRowsUpdatedAtOnce__c</fullName>
```
[the remark this answers is not in the document]
Comment (jamessimone) [reply rollup/main/default/objects/RollupControl__mdt/fields/MaxParentRowsUpdatedAtOnce__c.field-meta.xml:3]: See my response above - I've definitely already started unconsciously (as a result of working on this feature) incorporated what will eventually amount to taking you up on your stellar comment above. I don't want to further bloat this change by introducing a ton of other renamings (some of which are already noise in this PR introduced by the use of `apex-assist` helpfully pointing our where I was shadowing variable names).

File: rollup/main/default/classes/RollupTests.cls [hunk 5]
```diff
@@ -2646,7 +2646,7 @@ private class RollupTests {
     Rollup.maxQueryRowOverride = 1;
 
     Test.startTest();
-    Rollup.performFullRecalculation('Amount', 'AccountId', 'Id', 'AnnualRevenue', 'Account', 'Opportunity', 'SUM', null);
+    Rollup.performFullRecalculation('Amount', 'AccountId', 'Id', 'AnnualRevenue', 'Account', 'Opportunity', 'SUM', '');
```
[the remark this answers is not in the document]
Comment (jamessimone) [reply rollup/main/default/classes/RollupTests.cls:2649]: Ah, I can revert this - good catch. I had found that Apex has an issue with static methods that have been overloaded where the only difference with them (other than the last argument's type signature) was the visibility level of the method. So before I renamed one of them to `performFullRecalculationInner`, the compiler was complaining that `null` was ambiguous as an argument since there were now two methods that it could go towards, both named `performFullRecalculation`. In the end, I renamed one of them to include `inner` because while the above change made Apex happy, the LWC associated with the `@AuraEnabled` version of the method was somehow still able to "see" the private static method ... rendering the whole change here moot. I just forgot to back it out amidst the other work - great catch!

tl;dr - me monkey, smooth brain forgot

File: rollup/main/default/classes/RollupRelationshipFieldFinder.cls [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+  public class Traversal {
+    private Boolean isFinished = false;
+    private Boolean isAbortedEarly = false;
+
+    private final Map<Id, SObject> lookupIdToFinalRecords = new Map<Id, SObject>();
+    private Map<Id, List<Id>> lookupIdMap = new Map<Id, List<Id>>();
+    private final Map<Id, List<Id>> hierarchy = new Map<Id, List<Id>>();
+    private final RollupRelationshipFieldFinder finder;
+
+    public Traversal(RollupRelationshipFieldFinder finder) {
```
[the remark this answers is not in the document]
Comment (jamessimone) [reply rollup/main/default/classes/RollupRelationshipFieldFinder.cls:55]: none of the tests invoke the `Traversal` instructor (nor should they!) so this is an excellent call. consider it done!

File: rollup/main/default/classes/RollupRelationshipFieldFinder.cls [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+        return true;
+      }
+      CombinedHierarchy combinedHierarchy = this.getHierarchy(record, relationshipFieldName);
+      if (combinedHierarchy.currentHierarchy?.size() > 0 && combinedHierarchy.oldHierarchy?.size() > 0) {
+        // the last Ids present in the chain have to match, otherwise it's a reparenting
+        return combinedHierarchy.currentHierarchy[combinedHierarchy.currentHierarchy.size() - 1] !=
+          combinedHierarchy.oldHierarchy[combinedHierarchy.oldHierarchy.size() - 1];
+      } else {
+        // if there was only one hop, we can just compare the Ids
+        return currentLookupId != oldLookupId;
```
[the remark this answers is not in the document]
Comment (jamessimone) [reply rollup/main/default/classes/RollupRelationshipFieldFinder.cls:113]: I should update this note - it has to go last. The hierarchical comparison needs to be allowed to run first to avoid reporting false positives. Take a look at the comment block I made around line 235 for the full explanation!

File: rollup/main/default/classes/RollupCalculator.cls (hunk 1 above)
[the remark this answers is not in the document]
Comment (jamessimone) [reply rollup/main/default/classes/RollupCalculator.cls:684]: Update - tried the simple approach of reverting this change and updating the `handleDeleteConcat` method to do a replace on both the value _and_ the delimiter, but it didn't work. Since I'm not planning to spend more time on this tonight, I will plan to circle back around to this piece of feedback and clean that method up separately.

File: extra-tests/classes/RollupIntegrationTests.cls [hunk 8]
```diff
@@ -311,4 +311,54 @@ private class RollupIntegrationTests {
     System.assertEquals(secondChild.Name, updatedGreatGrandparent.Name, 'CONCAT_DISTINCT and reparenting should have worked');
     System.assertEquals(child.Name, updatedGreatGrandparentTwo.Name, 'CONCAT_DISTINCT and reparenting should have worked again');
   }
+
+  @isTest
+  static void shouldRunGrandparentRollupsWhenIntermediateObjectsAreUpdated() {
```
Comment (jongpie) [inline extra-tests/classes/RollupIntegrationTests.cls:316]: Niiiiiice

File: rollup/main/default/classes/Rollup.cls [hunk 9]
```diff
@@ -1534,11 +1557,40 @@ global without sharing virtual class Rollup implements Database.Batchable<SObjec
...
+      String currentLookup = (String) calcItem.get(fieldToken);
+      String oldLookup = (String) oldCalcItem.get(fieldToken);
+      if (currentLookup != oldLookup) {
+        wrapper.addRecordId(currentLookup);
+        wrapper.addRecordId(oldLookup);
+      }
+    }
+    return wrapper;
+  }
+
```
Comment (jongpie) [inline rollup/main/default/classes/Rollup.cls:1583]: It's only because you convinced me months ago to use apex-prettier that I feel obligated to point out... you have a few extra line breaks (line 1583 and 1587), you might want to run prettier on this class

File: rollup/main/default/classes/Rollup.cls [hunk 10]
```diff
@@ -1345,6 +1346,28 @@ global without sharing virtual class Rollup implements Database.Batchable<SObjec
           wrapper,
           rollupInfo.GrandparentRelationshipFieldPath__c
         );
+      } else if (
+        rollupInfo.CalcItem__c != calcItems.getSObjectType().getDescribe().getName() &&
+        String.isNotBlank(rollupInfo.GrandparentRelationshipFieldPath__c) &&
+        calcItems.isEmpty() == false
+      ) {
+        // it's a grandparent rollup triggered by an intermediate object
```
Comment (jongpie) [inline rollup/main/default/classes/Rollup.cls:1354]: It's beautiful

File: rollup/main/default/classes/Rollup.cls (hunk 9 above)
Comment (jamessimone) [reply rollup/main/default/classes/Rollup.cls:1586]: Damn! I just saw this comment after pushing a small cosmetic change and apex prettier did run (I’ve taken to trying to only run it right before committing). I think it bombs out some times though, honestly; it didn’t seem to pick up those extra line breaks. I will fix!

File: extra-tests/classes/RollupIntegrationTests.cls (hunk 8 above)
Comment (jamessimone) [reply extra-tests/classes/RollupIntegrationTests.cls:316]: So glad you tested that scenario
